sycl/doc/design/DeviceConfigFile.md

```diff
@@ -0,0 +1,332 @@
+compilation.
+
+Following, you can find the changes required in different parts of the project
+in more detail.
+
+### Changes to Build Infrastructure
+We need the information about the targets in multiple tools and compiler
+modules listed in [Requirements](#Requirements).  Thus, we need to make sure
+that the generation of the `.inc` file out of the `.td` file is done in time
+for all the consumers. The command we need to run for TableGen is `llvm-tblgen
+-gen-dynamic-tables -I /llvm-root/llvm/include/ input.td -o output.inc`.
+Additionally, we need to set dependencies adequately so that this command is
+run before any of the consumers need it.
+
+### Changes to the DPC++ Frontend
+To allow users to add new targets we provide a new flag:
+`fsycl-device-config-file=/path/to/file.yaml`. Users can pass a `.yaml` file
+describing the targets to be added/updated. An example of how such `.yaml` file
+should look like is shown below.
+```
+intel_gpu_skl:
+    aspects: [aspect_name1, aspect_name2]
+    may_support_other_aspects: true/false
+    sub-group-sizes: [1, 2, 4, 8]
+    aot-toolchain: ocloc
+    aot-toolchain-ocloc-device: skl
+```
+The frontend module should parse the user-provided `.yaml` file and update the
+map with the new information about targets. LLVM provides
+[YAML/IO](https://llvm.org/docs/YamlIO.html) library to easily parse `.yaml`
+files.
+
+As mentioned in [Requirements](#Requirements), there is an auto-detection
+mechanism for `aot-toolchain` and `aot-toolchain-%option_name` that is able to
+infer these from the target name. In the `.yaml` example shown above the target
+name is `intel_gpu_skl`. From that name, we can infer that `aot-toolchain` is
+`ocloc` because the name starts with `intel_gpu`. Also, we can infer that it needs
+`aot-toolchain-ocloc-device` set to `skl` just by keeping what is left after the
+prefix `intel_gpu`.
+
+#### Potential Issues/Limitations
+- Multiple targets with the same name: On the one hand, the compiler emits a
+warning so that the user is aware that multiple targets share the same name. On
+the other hand, it simply processes each new entry and updates the map with the
+latest information found.
+
+The auto-detection mechanism is a best effort to relieve users from specifying
+`aot-toolchain` and `aot-toolchain-%option_name` from well known devices. However, 
+it has its own limitations and potential issues:
+- Rules for target names: As of now, auto-detection is only available for Intel GPU
+targets. All targets starting with `intel_gpu_` will automatically set
+`aot-toolchain=ocloc` and `aot-toolchain-ocloc-device=suffix` being suffix the part
+left after `intel_gpu_` prefix.
+- User specifies `aot-toolchain` and `aot-toolchain-%option_name` for a target name 
+that can be auto-detected: user-specified information has precedence over auto-detected
+information.
+
+## Testing
+There is a danger that the device configuration file will get out-of-sync with the
```

Comment:
We have a test already for `any_device_has`, which goes over each item in `aspects.def` file and tries to instantiate `any_device_has` with that enumerator.

We could do the same for default device config. `if_architecture_is` could be used to detect which exact device we are running on and `.inc` file with default config can be included into SYCL unit-tests as well (perhaps we may want to select a different location for it, so it is clear that this test requires some compiler internals to be present), i.e. that is not exactly a runtime test.

However, I still don't think that such approach will allow us to be sure that nothing went out-of-sync, because it requires us to run the test on _every_ different target we have in our config file. We most likely won't have all of them available to us at any time even for manual run of the test, not to say about automated launch of it.

Not that we shouldn't run this test in our regular testing, but just highlighting that it won't automatically give us 100% coverage over all known targets.

An idea for an _extra_ test: we need a mechanism to compare list of aspects known to SYCL RT (`aspects.def`) with list of aspects available in `.td` file with default config description. This way, if someone adds a new aspect, they won't forget to register it in a `.td` file and hopefully, will not forget to add it to descriptions of known targets in that `.td` file.